sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -55,12 +55,25 @@ trait QueryPlanConstraints extends ConstraintHelper { self: LogicalPlan =>
 
 trait ConstraintHelper {
 
+  /**
+   * Infers an additional set of constraints from a given set of constraints.
+   */
+  def inferAdditionalConstraints(constraints: Set[Expression]): Set[Expression] = {
+    var inferred = inferEqualityConstraints(constraints)
+    var lastInequalityInferred = Set.empty[Expression]
+    do {
+      lastInequalityInferred = inferInequalityConstraints(constraints ++ inferred)
+      inferred ++= lastInequalityInferred
+    } while (lastInequalityInferred.nonEmpty)
```

Comment:
Could you hit a infinite loop with non deterministic filters? As they are never semantically equal to any other expression (including themselves). I hit that problem in #29650, where I was also working on constraint inference , but from `EqualNullSafe`.